backend/src/middleware/userMiddleware.spec.js

```diff
@@ -0,0 +1,81 @@
+import { GraphQLClient } from 'graphql-request'
+import { host } from '../jest/helpers'
+import Factory from '../seed/factories'
+
+const factory = Factory()
+let client
+
+afterAll(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('userMiddleware', () => {
+  describe('create User', () => {
+    const mutation = `
+      mutation($id: ID, $password: String!, $email: String!) {
+        CreateUser(id: $id, password: $password, email: $email) {
+          id
+        }
+      }
+    `
+    client = new GraphQLClient(host)
+
+    it('with password and email', async () => {
+      const variables = {
+        password: '123',
+        email: '123@123.de'
+      }
+      const expected = {
+        CreateUser: {
+          id: expect.any(String)
+        }
+      }
+      await expect(client.request(mutation, variables))
+        .resolves.toEqual(expected)
+    })
+  })
+
+  describe('update User', () => {
+    const mutation = `
+      mutation($id: ID!, $name: String) {
+        UpdateUser(id: $id, name: $name) {
+          name
+        }
+      }
+    `
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-02T17:52:38Z" title="Thursday, May 2nd 2019, 7:52:38 pm +02:00">May 2, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,364 @@
+import { GraphQLClient } from 'graphql-request'
+import { host } from '../jest/helpers'
+import Factory from '../seed/factories'
+
+const factory = Factory()
+let client
+
+afterAll(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('userMiddleware', () => {
+  describe('create User', () => {
+    const mutation = `
+      mutation($id: ID, $password: String!, $email: String!) {
+        CreateUser(id: $id, password: $password, email: $email) {
+          id
+        }
+      }
+    `
+    client = new GraphQLClient(host)
+
+    it('with password and email', async () => {
+      const variables = {
+        password: '123',
+        email: '123@123.de'
+      }
+      const expected = {
+        CreateUser: {
+          id: expect.any(String)
+        }
+      }
+      await expect(client.request(mutation, variables))
+        .resolves.toEqual(expected)
+    })
+
+    it('with ID, email and password', async () => {
+      const variables = {
+        password: '123',
+        id: 'u1',
+        email: '123@123.de'
+      }
+      const expected = {
+        CreateUser: {
+          id: 'u1'
```


I feel it's questionable to let the user choose the id. I know it's possible because the `ID` is in the `type User { ... }` but I think we should not support that users can choose their own id.

How about:
```js
    it('with ID, email and password', async () => {
      const variables = {
        password: '123',
        id: 'u1',
        email: '123@123.de'
      }
      const expected = 'ID cannot be given'
      await expect(client.request(mutation, variables))
        .rejects.toThrow(expected)
    })
```

Reply:
> [<img alt="ulfgebhardt" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/ulfgebhardt) **Authored by [ulfgebhardt](https://github.com/ulfgebhardt)**
_<time datetime="2019-05-06T11:01:05Z" title="Monday, May 6th 2019, 1:01:05 pm +02:00">May 6, 2019</time>_

---


I agree that might not be wise

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-05-06T17:56:15Z" title="Monday, May 6th 2019, 7:56:15 pm +02:00">May 6, 2019</time>_

---


Well then do one of the following:
1. Delete this test case or
2. Replace this test case with the code I suggested

Tests are executable specification. If you write a test case in which you pass the id to the mutation and expect that the server accepts it, this means that this is part of the expected behaviour. If you think this is not wise, then don't write the specification.

Reply:
> [<img alt="ulfgebhardt" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/ulfgebhardt) **Authored by [ulfgebhardt](https://github.com/ulfgebhardt)**
_<time datetime="2019-05-07T12:01:50Z" title="Tuesday, May 7th 2019, 2:01:50 pm +02:00">May 7, 2019</time>_

---


Deleted the testcase (comment) and added todo in order to recheck